uniffi_bindgen/src/lib.rs

```diff
@@ -157,6 +158,49 @@ pub trait BindingGenerator: Sized {
     ) -> Result<()>;
 }
 
+pub trait CodeOracle {
```

Comment:
Part of the motivation for my previous comment is that I really don't like trying to have a `CodeOracle` outside the bindings - it implies there will be a single trait that will capture the requirements of all internal and external binding languages, both current and future, which is something I'm skeptical of. Something that could "walk" the CI might mean the Oracles can remain in the bindings (or indeed, they can implement what they need any way they see fit).

(We saw something similar in the early days - trying to share too much code between the bindings ended up meaning we were forever working around impedance mismatches and avoiding this had significant benefits overall - eg, the `AsCodeType` trait was in a central place, but that turned out to be a mistake)

Reply:
Thanks! This makes a lot of sense! I continued with the work because I wanted it to pass first, so to be sure the Rust functionality can cover the base cases for the Python templating. I am at a point now where I can rework it, and take inspiration from your mentioned "[external type converter"](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_meta/src/group.rs#L88). 

So for this change, I before passing the generated `CI` down to the `PythonWrapper` (for example), we can walk through each field of the `ci`, adapt the names, and create a specific `python_ci`. Is this what you had in mind?

Reply:
I was thinking code in each of the bindings can explicitly do that itself. I'm actually a little vague because it's not immediately clear what will work! That "external type converter" isn't a great fit - I more meant very vaguely how it visits all parts of the metadata rather than the concrete implementation - for one thing it's walking a different data structure.